executor/executor.go

```diff
@@ -205,20 +214,35 @@ func Run(path string, env []string, emitter screwdriver.Emitter, build screwdriv
 		// Generate guid for the step
 		guid := uuid.NewV4().String()
 
+		completed := make(chan error, 1)
+
 		// Set current running step in emitter
 		emitter.StartCmd(cmd)
 		fmt.Fprintf(emitter, "$ %s\n", cmd.Cmd)
 
 		fReader := bufio.NewReader(f)
 
 		code, cmdErr = doRunCommand(guid, stepFilePath, emitter, f, fReader)
+		completed <- cmdErr
 
-		if err := api.UpdateStepStop(buildID, cmd.Name, code); err != nil {
-			return fmt.Errorf("Updating step stop %q: %v", cmd.Name, err)
+		select {
+		case <- completed:
+			if (firstError == nil) {
+				firstError = cmdErr
+			}
+		case buildTimeout := <- invokeTimeout:
```

Comment:
This is only for userCommands? What happens if teardown step is the one that takes a long time? We should still stop the build right? Can we refactor to use this for both.

Reply:
Considering post-build time (uploading artifacts) is influenced by the number of files generated in actual build operations it should be covered by the time out. In fact timeout should cover all, pre build, build steps, and post build

Reply:
> Considering post-build time (uploading artifacts) is influenced by the number of files generated in actual build operations it should be covered by the time out. In fact timeout should cover all, pre build, build steps, and post build, there is no need for separate settings for each steps.

That opens up the possibility that when a test times out, there are no artifacts saved. This would make debugging a hanging build very difficult if the artifacts are the only way to investigate the issue.

Reply:
That's a valid concern, lets start with timeout covering pre & user build steps, then we can introduce a separate one for post builds steps.

Reply:
I still believe Dao's concern is legitimate. I'll add it to the feature issue (https://github.com/screwdriver-cd/screwdriver/issues/545) and we can create a discussion around that piece or at least track the decisions there.

Reply:
Added it: https://github.com/screwdriver-cd/screwdriver/issues/545#issuecomment-373210885